### DISCUSSION TITLE
Fetching Data in Next.js: API vs direct database call

### DISCUSSION THREAD

**Discussion by taylorhlabs** [prisma/prisma#4897] [Q&A] [answered]
I am fairly new to React and Next.js so this is a question about best practices. 

Both of the following functions work to fetch the data I need. The first one calls a file in my Next.js pages/api folder, which then gets the data using Prisma. And the second one gets the data directly from the database using Prisma. Is one better than the other from a performance and security standpoint? If I don't need the reusability provided by putting the business logic into an API function (the first one), I am thinking that the second function is the best way to go given that it's easier to implement and doesn't require an extra API call. Is this an OK approach throughout my app? And any other "best practice" input would be appreciated. Thanks!

```ts
// Option 1: Calls API, then API uses Prisma to get data from database
export const getServerSideProps = async (context) => {
  const res = await fetch(`http://localhost:3001/api/resource/${context.params.id}`);
  const resource = await res.json();
  return { props: { ...resource } };
};
```
```ts
// Option 2: Uses Prisma to get data directly from database
export const getServerSideProps = async (context) => {
  const resource = await prisma.resource.findUnique({
    where: {
      id: parseInt(context.params.id)
    }
  });
  return { props: { ...resource } };
};
```

**Comment by ryands17** [comment, marked as answer]
Hey @taylorhlabs 👋 

As you are just fetching data, I would personally suggest using option 2 in which you can directly fetch data via Prisma as that would be better than creating an API specifically just for fetching data. As this will happen server-side, this will also not be included in your bundle so the frontend will stay slim and also you would avoid the latency of the API call.

You would need to use API's to update data though as that would be happening Client side. Let me know if that helps :)

**Reply by taylorhlabs** [reply]
Hey @ryands17 - thanks for the quick response. Makes sense 👍  

I am running into an issue with option 2 related to json parsing. If the returned object contains DATETIME fields from MySQL, I get an error.
```
Error: Error serializing `.resources[0].createdAt` returned from `getServerSideProps` in "/".
Reason: `object` ("[object Date]") cannot be serialized as JSON. Please only return JSON serializable data types.
```

If the object contains any DATETIME fields, do I need to use the API option? Or is there a workaround to be able to use Prisma directly?

**Reply by ryands17** [reply]
@taylorhlabs yes there is a workaround: https://github.com/blitz-js/superjson

Next.js cannot serialize dates by default so you would need to use this plugin that would automatically do this for you :)

**Reply by taylorhlabs** [reply]
OK, cool. And the reason that Option 1 works above without the plugin is because the data is being parsed by the server first before giving it back to Next.js?

**Reply by taylorhlabs** [reply]
To clarify, I can use Prisma inside the API method and I don't get the error but I get an error when trying to use Prisma within the `getServerSideProps` function.

**Reply by ryands17** [reply]
Yes you're absolutely correct. This is because the API will send you JSON and using Prisma directly will give you the actual date.

**Reply by taylorhlabs** [reply]
Thanks again for the input. After a bit more testing and playing around, I ended up sticking with the Prisma option and cleaning up the data after I get it back.
```
// get all resources
export async function getServerSideProps() {
  const resources = await prisma.resource.findMany();
  // fix dates. Next.js cannot parse DateTime on front end
  for (let resource of resources) {
    fixDates(resource);
  }
  return {
    props: { resources }
  };
}
```
```
// get resource by id
export async function getServerSideProps(context) {
  const resource = await prisma.resource.findUnique({
    where: {
      id: parseInt(context.params.id)
    }
  });
  // fix dates. Next.js cannot parse DateTime on front end
  fixDates(resource);
  return { props: resource };
}
```
```
// convert dates to strings to avoid json parse errors
function fixDates(obj) {
  obj.createdAt = obj.createdAt.toString();
  obj.updatedAt = obj.updatedAt.toString();
}
```

**Comment by Danimoz** [comment]
> Hey @taylorhlabs 👋
> 
> As you are just fetching data, I would personally suggest using option 2 in which you can directly fetch data via Prisma as that would be better than creating an API specifically just for fetching data. As this will happen server-side, this will also not be included in your bundle so the frontend will stay slim and also you would avoid the latency of the API call.
> 
> You would need to use API's to update data though as that would be happening Client side. Let me know if that helps :)

Can option2 be cached so as to avoid calling the database everytime. If yes? how can this be done?

**Reply by kayzenjuice** [reply]
Unfortunately, this is still an experimental feature in Next.js 13 using the app router. I believe they are working on making it stable.

If you want to cache a direct database call, you can use unstable_cache provided by Next.js. With experimental server actions turned on, you can also set tags to revalidate data by tag, as shown in the example below:

[Unstable Cache Docs Next Js 13](https://nextjs.org/docs/app/building-your-application/caching#unstable_cache)

```js

import { unstable_cache } from "next/cache";

const cachedNotes = await unstable_cache(
  async (searchTerm) => {
    let notes = [];
    if (searchTerm) {
      notes = await prisma.note.findMany({
        where: {
          title: {
            contains: searchTerm,
            mode: "insensitive",
          },
        },
      });
    } else {
      notes = await prisma.note.findMany();
    }

    return notes;
  },
  ["allNotes_search"],
  {
    tags: ["allnotes"],
    revalidate: 5,
  },
);

```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
